### ISSUE TITLE
Can't seem to get links to work...

### ISSUE THREAD

**Issue by hismightiness** [ActiveForums/ActiveForums#437]
I've tried every way I can find so far, but I'm unable to get links to work as expected on the DNNCommunity.org website.  If you create a new post and add a link, it's correctly linked.  When you view the post, the link is removed, but the URL remains.  If you edit the post, it's still there and linked as expected. 

Has anyone run into this?  Do you know what the fix is?

**Comment by hismightiness** [comment]
Uh...  All of a sudden ONE of the links seems to be working on the first post I've noticed so far!  🤔 

https://dnncommunity.org/forums/aft/39

What are the specific steps anyone here have used to make links work consistently across all forums?

**Comment by OllyHodgson** [comment]
Today I've made a few posts on the DNNCommunity.org forums.

1. If I just paste in a URL, it's not linked.
2. If I hit the link button in the toolbar, and paste in a URL, it's linked in the editor, but this does not show up as a link in the submitted forum post.
3. If I type in some text, select it, then hit the link button in the toolbar, and paste in a URL, it's linked up and works perfectly.

If I follow step 2, and then quote the post, the link shows up again in the quoted post when editing, but not when submitted to the site. So it seems like there's some pre-render filtering, which takes out any link where the text and the href are the same.

**Comment by Timo-Breumelhof** [comment]
I switched filters off for one of the (MVP) forums and that does not make a difference

**Comment by Timo-Breumelhof** [comment]
Noting here it seems: https://github.com/ActiveForums/ActiveForums/blob/592f0f76c807f25615902be1dc4b25886d621243/CustomControls/Views/TopicView.cs

**Comment by Timo-Breumelhof** [comment]
Ok, switching off auto links helped. So there's an issue with Autolinks here:https://github.com/ActiveForums/ActiveForums/blob/592f0f76c807f25615902be1dc4b25886d621243/class/Utilities.cs

The regex looks for http links not https links, should be an easy fix..

**Comment by Timo-Breumelhof** [comment]
FYI, when you switch AutoLinks on this one works:

<a href="https://www.40fingers.net/" target="_blank">Link test 1</a>
This one doesn't
<a href="https://www.40fingers.net/" target="_blank">https://www.40fingers.net/</a>
And this one does:
<a href="http://www.40fingers.net/" target="_blank">http://www.40fingers.net/</a>

**Comment by Timo-Breumelhof** [comment]
BTW, this seems to fail too: 
<a href="https://www.40fingers.net/" target="_blank">Link test 1</a><a href="https://www.40fingers.net/" target="_blank">https://www.40fingers.net/</a> (no space in between the links)

**Comment by Timo-Breumelhof** [comment]
I think I fixed it.
I tested the function on https://dotnetfiddle.net/ and AFAIK it works the way it should like this:

using System;
using System.Text.RegularExpressions;
using System.Web;
					
public class Program
{
	public static void Main()
	{
		
		AutoLinks("https://www.40fingers.net/" , "") ;
		AutoLinks("<a href='https://www.40fingers.net/' target='_blank'>Link test 1</a>" , "");
		AutoLinks("<a href='https://www.40fingers.net/' target='_blank'>https://www.40fingers.net/</a>" , "");
		AutoLinks("<a href='https://www.40fingers.net/>https://www.40fingers.net/</a><a href='https://www.40fingers.net/>https://www.40fingers.net/</a>" , "");
		
	}
	
	public static string AutoLinks(string text, string currentSite)
        {
            var original = text;
            if (!(string.IsNullOrEmpty(text)))
            {
                const string encodedHref = "&lt;a.*?href=[\"'](?<url>.*?)[\"'].*?&gt;(http[s]?.*?)&lt;/a&gt;"; // Encoded href regex
                const string regHref = "<a.*?href=[\"'](?<url>.*?)[\"'].*?>(?<http>http[s]?.*?)</a>"; // Regular href regex

				Console.WriteLine("Start:" + text );
				// Replace encoded url with decoded url
                foreach (Match m in Regex.Matches(text, encodedHref, RegexOptions.IgnoreCase))
                    text = text.Replace(m.Value, HttpUtility.HtmlDecode(m.Value));

				Console.WriteLine("s2:" + text );


				Console.WriteLine("s3:" + text );
				
				
				
                if (string.IsNullOrEmpty(text))
                {
                    return original;
                }
				

					
				String strRegexUrl = @"(?<!['""]+|[>]+)http[s]?://([\w+?\.\w+])+([a-zA-Z0-9\~\!\@\\#\$\%\^\&amp;\*\(\)_\-\=\+\\\/\?\.\:\;\'\,]*)?";
				

                text = Regex.Replace(text, strRegexUrl, m => ReplaceLink(m, currentSite, text), RegexOptions.IgnoreCase);
				
				
				
                if (string.IsNullOrEmpty(text))
                {
                    return original;
                }
            }
		
		Console.WriteLine("End:" + text + "\n");
		
            return text;
        }
	
	        private static string ReplaceLink(Match match, string currentSite, string text)
        { 
            const int maxLengthAutoLinkLabel = 47;
            const string outSite = "<a href=\"{0}\" target=\"_blank\" rel=\"nofollow\">{1}</a>";
            const string inSite = "<a href=\"{0}\">{1}</a>";
            var url = match.Value;
            if (url.ToLowerInvariant().Contains("jpg") || url.ToLowerInvariant().Contains("gif") || url.ToLowerInvariant().Contains("png") || url.ToLowerInvariant().Contains("jpeg"))
                return url;

            //
            // Ignore it when there is a preceeding a or img.
            //
            var xStart = 0;
            if ((match.Index - 10) > 0)
                xStart = match.Index - 10;
            
            if (text.Substring(xStart, 10).ToLowerInvariant().Contains("href"))
                return url;

            if (text.Substring(xStart, 10).ToLowerInvariant().Contains("src"))
                return url;

            if (text.Substring(xStart, 10).ToLowerInvariant().Contains("="))
                return url;
        
            var urlText = match.Value;
            if (urlText.Length > maxLengthAutoLinkLabel)
                urlText = match.Value.Substring(0, maxLengthAutoLinkLabel - 22) + "..." + match.Value.Substring(match.Value.Length - 20);

            return url.ToLowerInvariant().Contains(currentSite.ToLowerInvariant()) ? string.Format(inSite, url, urlText) : string.Format(outSite, url, urlText);
}

	
}

**Comment by Timo-Breumelhof** [comment]
FYI, 
A. I changed the Regex's to include https, maybe one should aIso add FTP? https://github.com/ActiveForums/ActiveForums/blob/592f0f76c807f25615902be1dc4b25886d621243/class/Utilities.cs#L427

B. I added a negative lookbehind to the check for urls to "autolink". https://github.com/ActiveForums/ActiveForums/blob/592f0f76c807f25615902be1dc4b25886d621243/class/Utilities.cs#L441
 This means that the code that removes exiting links is obsolete now
https://github.com/ActiveForums/ActiveForums/blob/592f0f76c807f25615902be1dc4b25886d621243/class/Utilities.cs#L433

FYI, this lookbehind will fail is someone insert a quote before the url BTW (in that the url will not be auto injected)
